contrib/pam_zfs_key/pam_zfs_key.c

```diff
@@ -443,9 +475,56 @@ zfs_key_config_free(zfs_key_config_t *config)
 	free(config->homes_prefix);
 }
 
+static int
+find_dsname_by_prop_value(zfs_handle_t *zhp, void *data)
+{
+	zfs_key_config_t *target = data;
+	zfs_type_t type = zfs_get_type(zhp);
+	char mountpoint[ZFS_MAXPROPLEN];
+	char *dsname = NULL;
+
+	// Iterate over any nested datasets.
+	if (zfs_iter_filesystems(zhp, find_dsname_by_prop_value, data) != 0) {
+		zfs_close(zhp);
+		return (1);
+	}
+
+	// Skip any datasets whose type does not match.
+	if ((type & ZFS_TYPE_FILESYSTEM) == 0) {
+		zfs_close(zhp);
+		return (0);
+	}
+
+	dsname = strdup(zfs_get_name(zhp));
+	if (target->homes_prefix != NULL) {
+		if (strncmp(dsname, target->homes_prefix,
+		    strlen(target->homes_prefix)) != 0) {
+			zfs_close(zhp);
+			return (0);
+		}
+	}
+
+	if (target->homedir != NULL) {
+		(void) zfs_prop_get(zhp, ZFS_PROP_MOUNTPOINT, mountpoint,
+		    sizeof (mountpoint), NULL, NULL, 0, B_FALSE);
+		if (strcmp(target->homedir, mountpoint) != 0) {
+			zfs_close(zhp);
```

Comment:
`dsname` is leaked.

Reply:
will `free(dsname)` in next revision, thanks.